app/models/user.rb

```diff
@@ -725,6 +725,28 @@ def update_ip_address!(new_ip_address)
     unless ip_address == new_ip_address || new_ip_address.blank?
       update_column(:ip_address, new_ip_address)
     end
+
+    if SiteSetting.keep_old_ip_address_count > 0
+      DB.exec(<<~SQL, user_id: self.id, ip_address: new_ip_address, current_timestamp: Time.zone.now)
+      INSERT INTO user_ip_address_histories (user_id, ip_address, created_at, updated_at)
+      VALUES (:user_id, :ip_address, :current_timestamp, :current_timestamp)
+      ON CONFLICT (user_id, ip_address)
+      DO
+        UPDATE SET updated_at = :current_timestamp
+      SQL
+
+      DB.exec(<<~SQL, user_id: self.id, offset: SiteSetting.keep_old_ip_address_count)
+      DELETE FROM user_ip_address_histories
+      WHERE id IN (
+        SELECT
+          id
+        FROM user_ip_address_histories
+        WHERE user_id = :user_id
+        ORDER BY updated_at DESC
+        OFFSET :offset
+      )
+      SQL
```

Comment:
This will be evaluated with every request so INSERTING and then DELETING additional may be suboptimal.

Maybe we should use CASE statement to INSERT a new record (if still in keep_old_ip_address_count limit) or UPDATE the oldest record.

Problem with that is that if you set keep_old_ip_address_count to 10 and then change it to 5, you will still have 10 records in DB and it will be never cleaned. So I am not sure if that would be much better in the end

Reply:
@lis2 Actually, this was a mistake I made, I changed it so that we only log to the `user_ip_address_histories` table only if the IP address has been changed.